Distributed tracing 2.9 (#62410)
openshift/openshift-docs: modules/distr-tracing-tempo-install-cli.adoc

// Module included in the following assemblies:
//
//* distr_tracing_tempo/distr-tracing-tempo-installing.adoc

:_content-type: PROCEDURE
[id="distr-tracing-tempo-install-cli_{context}"]
= Installing {TempoName} by using the CLI

> 

Author: ```suggestion

Installing {TempoName} by using the CLI involves installing the Tempo Operator, creating a Secret and namespace, and using the Operator to deploy a TempoStack instance in the namespace for use with the Secret.
//|Tempo Operator| might need changing to {TempoShortName}.

```

> 

Author: ```suggestion

Installing {TempoName} by using the CLI involves installing the Tempo Operator, creating a Secret and namespace, and using the Operator to deploy a TempoStack instance in the namespace for use with the Secret.
//|Tempo Operator| might need changing to {TempoShortName}.

```

+
.TempoStack CR for AWS S3 and MinIO storage
====
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
  namespace: tempostacknamespace
spec:
  storageSize: 1Gi
  storage:
    secret:
      name: minio-test
      type: s3
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
The stack deployed in this example is configured to receive Jaeger Thrift over HTTP and OpenTelemetry Protocol (OTLP), which permits visualizing the data with the Jaeger UI.
====

.. Apply the customized CR.
+
[source,terminal]
----
$ oc apply -f - << EOF
<TempoStack_custom_resource>
EOF
----

.WIP:

. Command(s) to verify successful deployment of the TempoStack instance
+
For example:
.. A command that prints output to verify that pods are created on the namespace.
.. A command that enables the CLI user to watch for the stack to stabilize.

. If applicable, a command to open any UI console

> . Any other commands?

Author: @rubenvp8510 and @IshwarKanse, when you can find the time, could you come up with suggested changes for these lines here?

:_content-type: PROCEDURE
[id="distr-tracing-tempo-install-cli_{context}"]
= Installing the {TempoShortName} by using the CLI

You can install the {TempoShortName} on the command line.

.Prerequisites

* An active OpenShift CLI (`oc`) session by a xref:https://docs.openshift.com/container-platform/latest/admin_solutions/user_role_mgmt.html#creating-a-cluster-administrator[cluster administrator with the `cluster-admin` role]. See link:https://docs.openshift.com/container-platform/{ocp4-ver}/cli_reference/openshift_cli/getting-started-cli.html[Getting started with the OpenShift CLI].
+
TIP: Ensure that your OpenShift CLI (`oc`) version is up to date and matches your {product-title} version.

* You are using a supported provider of object storage: link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[Red Hat OpenShift Data Foundation], link:https://min.io/[MinIO], link:https://aws.amazon.com/s3/[Amazon S3], link:https://azure.microsoft.com/en-us/products/storage/blobs/[Azure Blob Storage], link:https://cloud.google.com/storage/[Google Cloud Storage].

. Install the {TempoOperator}:

.. Create a subscription.
+
[source,yaml]
----

> $ oc apply -f - << EOF

Reviewer: For Tempo operator installation, we can use the same project that we are using for Tempo operator from UI. 

```
apiVersion: project.openshift.io/v1
kind: Project
metadata:
  labels:
    kubernetes.io/metadata.name: openshift-tempo-operator
    openshift.io/cluster-monitoring: "true"
  name: openshift-tempo-operator

---
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: openshift-tempo-operator-9bb49
  namespace: openshift-tempo-operator
spec:
  upgradeStrategy: Default

---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: tempo-product
  namespace: openshift-tempo-operator
spec:
  channel: stable
  installPlanApproval: Automatic
  name: tempo-product
  source: redhat-operators
  sourceNamespace: openshift-marketplace
```

Author: Done.

apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: tempo-product
  namespace: openshift-operators
spec:
  channel: stable
  installPlanApproval: Automatic
  name: tempo-product
  source: redhat-operators
  sourceNamespace: openshift-marketplace
EOF
----

.. Check the Operator status.
+
[source,terminal]
----
$ oc get csv -n openshift-operators
----

. Run `oc apply -f` to create a Secret for your object storage bucket.
+
[TIP]
====
[source,terminal]
----
$ oc apply -f <secret_file>.yaml
----

[source,terminal]
----
$ oc apply -f - << EOF
<object_storage_secret>
EOF
----
====
+
--
include::snippets/distr-tracing-tempo-required-secret-parameters.adoc[]
--
+
--
include::snippets/distr-tracing-tempo-secret-example.adoc[]
--

. Use `oc create namespace` or `oc apply` to create a namespace for the *TempoStack* instance that you will create in the next step.
+
[TIP]
====
[source,terminal]
----
$ oc create namespace <namespace_value_of_your_choice>
----

[source,terminal]
----

> $ oc apply -f - << EOF

Reviewer: Remove this config to apply monitoring labels for the namespace in which TempoStack is created. This labels are needed only in the namespace in which Tempo Operator is installed and not in the namespace in which TempoStack is created, for TempoStack namespace, we use user workload monitoring.

Author: Done. Thank you for the explanation.

====
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
  namespace: tempostacknamespace
spec:
  storageSize: 1Gi
  storage:
    secret:
      name: minio-test
      type: s3
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
The stack deployed in this example is configured to receive Jaeger Thrift over HTTP and OpenTelemetry Protocol (OTLP), which permits visualizing the data with the Jaeger UI.
====

.. Apply the customized CR.
+
[source,terminal]
----
$ oc apply -f - << EOF
<TempoStack_custom_resource>
EOF
----

link:https://github.com/openshift/openshift-docs/pull/62410/files#r1284356025[Waiting for SME input for the following:]

* Command(s) to verify successful deployment of the TempoStack instance
+
For example:
** A command that prints output to verify that pods are created on the namespace.
** A command that enables the CLI user to watch for the stack to stabilize.

* If applicable, a command to open any UI console

> * Any other commands?

Author: @rubenvp8510 and @IshwarKanse, when you can find the time, could you come up with suggested changes for these lines here?

Reviewer: For checking the TempoStack status, we can use these commands.
1. Make sure all the component status is running and the condition in status is in Ready state.
```
$ oc get tempostacks.tempo.grafana.com simplest -o yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  creationTimestamp: "2023-08-07T07:46:48Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: tempo-operator
    tempo.grafana.com/distribution: openshift
  name: simplest
  namespace: tracing-system
  resourceVersion: "60486"
  uid: 4f68828e-8233-4d50-9885-96b663b030f7
spec:
  images:
.............REDACTED ......
  limits:
    global:
      ingestion: {}
      query:
        maxSearchDuration: 0s
  managementState: Managed
  observability:
    metrics: {}
    tracing:
      jaeger_agent_endpoint: localhost:6831
  replicationFactor: 1
  resources:
    total:
      limits:
        cpu: "2"
        memory: 2Gi
  retention:
    global:
      traces: 48h0m0s
  search:
    defaultResultLimit: 20
    maxDuration: 0s
  serviceAccount: tempo-simplest
  storage:
    secret:
      name: minio-secret
      type: s3
  storageSize: 20Gi
  template:
    compactor: {}
    distributor:
      replicas: 1
    gateway:
      component: {}
      enabled: false
      ingress:
        route: {}
    ingester:
      replicas: 1
    querier: {}
    queryFrontend:
      component: {}
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
status:
  components:
    compactor:
      Running:
      - tempo-simplest-compactor-77cc4b9c9f-7grk4
    distributor:
      Running:
      - tempo-simplest-distributor-6759b88d99-wdqfn
    ingester:
      Running:
      - tempo-simplest-ingester-0
    querier:
      Running:
      - tempo-simplest-querier-74d77fc548-f4nxd
    queryFrontend:
      Running:
      - tempo-simplest-query-frontend-54b7c465b5-fdrxs
  conditions:
  - lastTransitionTime: "2023-08-07T07:47:12Z"
    message: Some TempoStack components are pending on dependencies
    reason: PendingComponents
    status: "False"
    type: Pending
  - lastTransitionTime: "2023-08-07T07:47:12Z"
    message: All components are operational
    reason: Ready
    status: "True"
    type: Ready
  operatorVersion: 0.3.0
  tempoVersion: 2.1.1
```
2. Make sure all the component pods are running.
```
$ oc get pods
NAME                                             READY   STATUS    RESTARTS   AGE
tempo-simplest-compactor-77cc4b9c9f-7grk4        1/1     Running   0          3m1s
tempo-simplest-distributor-6759b88d99-wdqfn      1/1     Running   0          3m1s
tempo-simplest-ingester-0                        1/1     Running   0          3m1s
tempo-simplest-querier-74d77fc548-f4nxd          1/1     Running   0          3m1s
tempo-simplest-query-frontend-54b7c465b5-fdrxs   2/2     Running   0          3m1s
```
To get the route for the TempoStack query UI, run the following command.
```
$ oc get route -n tempostacknamespace
NAME                            HOST/PORT                                                                                     PATH   SERVICES                        PORT        TERMINATION   WILDCARD
tempo-simplest-query-frontend   tempo-simplest-query-frontend-tempostacknamespace.apps.mytest-55.qe.gcp.mycluster.test.com          tempo-simplest-query-frontend   jaeger-ui   edge          None

```

// Module included in the following assemblies:
//
//* distr_tracing_tempo/distr-tracing-tempo-installing.adoc

:_content-type: PROCEDURE
[id="distr-tracing-tempo-install-cli_{context}"]
= Installing the {TempoShortName} by using the CLI

You can install the {TempoShortName} on the command line.

.Prerequisites

* An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.
+
TIP: Ensure that your OpenShift CLI (`oc`) version is up to date and matches your {product-title} version.

> * You are using a supported provider of object storage: link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[Red Hat OpenShift Data Foundation], link:https://min.io/[MinIO], link:https://aws.amazon.com/s3/[Amazon S3], link:https://azure.microsoft.com/en-us/products/storage/blobs/[Azure Blob Storage], link:https://cloud.google.com/storage/[Google Cloud Storage].

Author: Thank you for the review. I'll probably keep the original to avoid confusion about whether the type of object storage is supported or the provider.

> 

Author: @shipsing, thank you so much for catching this!

. Install the {TempoOperator}:

> .. Create a subscription.

Author: Thank you for the suggestion. It's not clear to me how I'd convince the reader that it's something they should do for this project.

    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
<1> The value of `name` in the `metadata` of the Secret.
<2> The accepted values are `azure` for Azure Blob Storage, `gcs` for Google Cloud Storage, and `s3` for Amazon S3 or MinIO or Red Hat OpenShift Data Foundation.
+
.TempoStack CR for AWS S3 and MinIO storage
====
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
  namespace: tempostacknamespace
spec:
  storageSize: 1Gi
  storage:
    secret:
      name: minio-test
      type: s3
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
The stack deployed in this example is configured to receive Jaeger Thrift over HTTP and OpenTelemetry Protocol (OTLP), which permits visualizing the data with the Jaeger UI.
====

.. Apply the customized CR.
+
[source,terminal]
----
$ oc apply -f - << EOF
<TempoStack_custom_resource>
EOF
----

.Verification

> . Verify that the `status` of all TempoStack `components` is `Running` and the `conditions` are `type: Ready`.

Author: Thank you.

  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
<1> The value of `name` in the `metadata` of the Secret.
<2> The accepted values are `azure` for Azure Blob Storage, `gcs` for Google Cloud Storage, and `s3` for Amazon S3 or MinIO or Red Hat OpenShift Data Foundation.
+
.TempoStack CR for AWS S3 and MinIO storage
====
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
  namespace: tempostacknamespace
spec:
  storageSize: 1Gi
  storage:
    secret:
      name: minio-test
      type: s3
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
The stack deployed in this example is configured to receive Jaeger Thrift over HTTP and OpenTelemetry Protocol (OTLP), which permits visualizing the data with the Jaeger UI.
====

.. Apply the customized CR.
+
[source,terminal]
----
$ oc apply -f - << EOF
<TempoStack_custom_resource>
EOF
----

> link:https://github.com/openshift/openshift-docs/pull/62410/files#r1284564654[Waiting for SME input for the following:]

Author: ```suggestion
```

EOF
----

.. Check the Operator status.
+
[source,terminal]
----
$ oc get csv -n openshift-tempo-operator
----

. Run `oc apply -f` to create a Secret for your object storage bucket.
+
[TIP]
====
[source,terminal]
----
$ oc apply -f <secret_file>.yaml
----

[source,terminal]
----
$ oc apply -f - << EOF
<object_storage_secret>
EOF
----
====
+
--
include::snippets/distr-tracing-tempo-required-secret-parameters.adoc[]
--
+
--
include::snippets/distr-tracing-tempo-secret-example.adoc[]
--

. Run `oc new-project` to create an OpenShift project of your choice for the *TempoStack* instance that you will create in the next step.
+
[TIP]
====
[source,terminal]
----
$ oc new-project <openshift_project_of_tempostack_instance>
----

[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: project.openshift.io/v1
kind: Project
metadata:
  labels:
    kubernetes.io/metadata.name: <openshift_project_of_tempostack_instance>
    openshift.io/cluster-monitoring: "true"
  name: <openshift_project_of_tempostack_instance>
EOF
----
====

. Create a *TempoStack* instance in the OpenShift project that you created for the *TempoStack* instance in the previous step.
+

> NOTE: You can create multiple *TempoStack* instances in separate projects on the same cluster.

Reviewer 2: we can move this note out of this procedure

Author: Thank you for the suggestion. I already mentioned this possibility at the assembly level, so this is just an additional reminder.

// Module included in the following assemblies:
//
//* distr_tracing_tempo/distr-tracing-tempo-installing.adoc

:_content-type: PROCEDURE
[id="distr-tracing-tempo-install-cli_{context}"]
= Installing the {TempoShortName} by using the CLI

You can install the {TempoShortName} on the command line.

.Prerequisites

* An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.
+
[TIP]
====
* Ensure that your OpenShift CLI (`oc`) version is up to date and matches your {product-title} version.

* Run `oc login`:
+
[source,terminal]
----
$ oc login --username=<your_username> 
----
====

* You are using a supported provider of object storage: link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[Red Hat OpenShift Data Foundation], link:https://min.io/[MinIO], link:https://aws.amazon.com/s3/[Amazon S3], link:https://azure.microsoft.com/en-us/products/storage/blobs/[Azure Blob Storage], link:https://cloud.google.com/storage/[Google Cloud Storage].

> . Install the {TempoOperator}:

Author: ```suggestion
.Procedure

. Install the {TempoOperator}:
```

.. Create a subscription.
+
[source,yaml]
----
$ oc apply -f - << EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: tempo-product
  namespace: openshift-tempo-operator
spec:
  channel: stable
  installPlanApproval: Automatic
  name: tempo-product
  source: redhat-operators
  sourceNamespace: openshift-marketplace
EOF
----

.. Check the Operator status.
+
[source,terminal]
----
$ oc get csv -n openshift-tempo-operator
----

. Run `oc apply -f` to create a Secret for your object storage bucket.
+
[TIP]
====
[source,terminal]
----
$ oc apply -f <secret_file>.yaml
----

[source,terminal]
----
$ oc apply -f - << EOF
<object_storage_secret>
EOF
----
====
+
--
include::snippets/distr-tracing-tempo-required-secret-parameters.adoc[]
--
+
--
include::snippets/distr-tracing-tempo-secret-example.adoc[]
--

. Run `oc new-project` to create an OpenShift project of your choice for the *TempoStack* instance that you will create in the next step.
+
[TIP]
====
[source,terminal]
----
$ oc new-project <openshift_project_of_tempostack_instance>
----

[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: project.openshift.io/v1
kind: Project
metadata:

>   labels:
>     kubernetes.io/metadata.name: <openshift_project_of_tempostack_instance>
>     openshift.io/cluster-monitoring: "true"

Author: ```suggestion
```

> metadata:
>   labels:
>     kubernetes.io/metadata.name: <openshift_project_of_tempostack_instance>
>     openshift.io/cluster-monitoring: "true"
>   name: <openshift_project_of_tempostack_instance>

Author: ```suggestion
```

// Module included in the following assemblies:
//
//* distr_tracing_tempo/distr-tracing-tempo-installing.adoc

:_content-type: PROCEDURE
[id="distr-tracing-tempo-install-cli_{context}"]
= Installing the {TempoShortName} by using the CLI

> You can install the {TempoShortName} on the command line.

Reviewer 3: ```suggestion
You can install the {TempoShortName} from the command line.
```

.Prerequisites

> * An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.

Author: Thank you so much. I didn't know that. As you can imagine, I'm new to the OpenShift docs (since two months and working on this PR only so far). I clearly need to catch up on onboarding. Also, it didn't occur to me to study the attributes list: will do so after this release! Thank you again.

> * An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.

Reviewer 4: It's a relatively new attribute, it's not widely used because we haven't had time to sweep the repo for old instances yet. So, pretty easy to miss I think! :) I am super impressed at your first OCP docs PR - really legendary effort :D

> * An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.

Reviewer 5: +1 to what Jeana said; legendary effort indeed!!

> * An active OpenShift CLI (`oc`) session by a cluster administrator with the `cluster-admin` role.

Author: Fixed in [this commit](https://github.com/openshift/openshift-docs/pull/62410/commits/dec1098fb34d993481085d6344ebd52a53013e29).

+
[TIP]
====
* Ensure that your OpenShift CLI (`oc`) version is up to date and matches your {product-title} version.

* Run `oc login`:
+
[source,terminal]
----
$ oc login --username=<your_username>
----
====

* You are using a supported provider of object storage: link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[Red Hat OpenShift Data Foundation], link:https://min.io/[MinIO], link:https://aws.amazon.com/s3/[Amazon S3], link:https://azure.microsoft.com/en-us/products/storage/blobs/[Azure Blob Storage], link:https://cloud.google.com/storage/[Google Cloud Storage].

.Procedure

. Install the {TempoOperator}:

> .. Create an OpenShift project for the {TempoOperator}.

Reviewer 4: Standard method suggested below. As an alternative, you could put "by running the following commands:" at the end of step 1 and leaving the descriptions of each as you have them

```suggestion
.. Create an OpenShift project for the {TempoOperator} by running the following command:
```

Author: Thank you. Already applied globally.

+
[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: project.openshift.io/v1
kind: Project
metadata:
  labels:
    kubernetes.io/metadata.name: openshift-tempo-operator
    openshift.io/cluster-monitoring: "true"
  name: openshift-tempo-operator
EOF
----

.. Create an Operator group.
+
[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: openshift-tempo-operator
  namespace: openshift-tempo-operator
spec:
  upgradeStrategy: Default
EOF
----

.. Create a subscription.
+
[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: tempo-product
  namespace: openshift-tempo-operator
spec:
  channel: stable
  installPlanApproval: Automatic
  name: tempo-product
  source: redhat-operators
  sourceNamespace: openshift-marketplace
EOF
----

.. Check the Operator status.
+
[source,terminal]
----
$ oc get csv -n openshift-tempo-operator
----

> . Run `oc apply -f` for a YAML file or standard input to create a Secret for your object storage bucket:
> +
> [source,terminal]
> ----
> $ oc apply -f <secret_file>.yaml
> ----
> +
> [source,terminal]
> ----
> $ oc apply -f - << EOF
> <object_storage_secret>
> EOF
> ----

Reviewer 4: Is "Secret" an API object? Typically we use either "secret" or "`Secret` object", rather than treating it as a plain proper noun.

Command steps need "by running the following command:" at the end. Also cannot put two commands in one step. Possible approach:

```suggestion
. Create a secret for your object storage bucket by running one of the following commands:

** To create a secret from a YAML file:
+
[source,terminal]
----
$ oc apply -f <secret_file>.yaml
----

 ** To create a secret from standard input:
+
[source,terminal]
----
$ oc apply -f - << EOF
<object_storage_secret>
EOF
----
```

Author: Fantastic suggestion! Thank you. Applying manually.

+
--
include::snippets/distr-tracing-tempo-required-secret-parameters.adoc[]
--
+
--
include::snippets/distr-tracing-tempo-secret-example.adoc[]
--

> . Run `oc new-project` to create an OpenShift project of your choice for the *TempoStack* instance that you will create in the next step.
> +
> [TIP]
> ====
> [source,terminal]
> ----
> $ oc new-project <openshift_project_of_tempostack_instance>
> ----
> 
> [source,terminal]
> ----
> $ oc apply -f - << EOF
> apiVersion: project.openshift.io/v1
> kind: Project
> metadata:
>   name: <openshift_project_of_tempostack_instance>
> EOF
> ----
> ====

Reviewer 4: Each command needs an intro sentence explaining what the command does. In this case, I would suggest making these substeps of "Run `oc new-project` [...]" similar to the below. [Reviewer note: I don't actually know what action to say for the second substep, but you get the idea :slightly_smiling_face:]

```suggestion
. Run `oc new-project` to create an OpenShift project of your choice for the *TempoStack* instance that you will create.

.. Create a project by running the following command:
[source,terminal]
----
$ oc new-project <openshift_project_of_tempostack_instance>
----

.. Apply the configuration by running the following command:
[source,terminal]
----
$ oc apply -f - << EOF
apiVersion: project.openshift.io/v1
kind: Project
metadata:
  name: <openshift_project_of_tempostack_instance>
EOF
----
```

Author: I get what you're saying. I rewrote it based on my understanding of the context. Thank you so much.

. Create a *TempoStack* instance in the OpenShift project that you created for the *TempoStack* instance in the previous step.
+
NOTE: You can create multiple *TempoStack* instances in separate projects on the same cluster.
+
.. Customize the `TempoStack` custom resource (CR):
+
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: sample
  namespace: <openshift_project_of_tempostack_instance>
spec:
  storageSize: 1Gi
  storage:
      secret:
          name: <secret-name> <1>
          type: <secret-provider> <2>
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----

> <1> The value of the `name` in the `metadata` of the Secret.

Reviewer 4: Or "`Secret` object" (whichever you prefer) should be changed throughout.
```suggestion
<1> The value of the `name` in the `metadata` of the secret.
```

Author: Fixed throughout the PR.

> <2> The accepted values are `azure` for Azure Blob Storage, `gcs` for Google Cloud Storage, and `s3` for Amazon S3 or MinIO or Red Hat OpenShift Data Foundation.

Reviewer 4: Is `s3` for all three that follow it? Might consider an unordered list of these to help keep them straight. In any case, if it is for all three suggest the serial comma format:

```suggestion
<2> The accepted values are `azure` for Azure Blob Storage, `gcs` for Google Cloud Storage, and `s3` for Amazon S3, MinIO, or Red Hat OpenShift Data Foundation.
```

You could also use `;` since there is internal punctuation in one of the items:
```suggestion
<2> The accepted values are `azure` for Azure Blob Storage; `gcs` for Google Cloud Storage; and `s3` for Amazon S3, MinIO, or Red Hat OpenShift Data Foundation.
```

Author: Thank you so much. I've used your second suggestion with `;` and applied it manually.

+
.TempoStack CR for AWS S3 and MinIO storage
====
[source,yaml]
----
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
  namespace: openshift_project_of_tempostack_instance
spec:
  storageSize: 1Gi
  storage:
    secret:
      name: minio-test
      type: s3
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          route:
            termination: edge
          type: route
----
The stack deployed in this example is configured to receive Jaeger Thrift over HTTP and OpenTelemetry Protocol (OTLP), which permits visualizing the data with the Jaeger UI.
====

> .. Apply the customized CR.

Reviewer 4: ```suggestion
.. Apply the customized CR by running the following command:
```

Author: Thank you. Applied locally.

+
[source,terminal]
----
$ oc apply -f - << EOF
<TempoStack_custom_resource>
EOF
----

> .Verification

Reviewer 4: Some command steps under this are missing `[source,terminal]`, all are missing "by running the following command:"

Author: Thank you for spotting. Fixed.

apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: tempo-product
  namespace: openshift-tempo-operator
spec:
  channel: stable
  installPlanApproval: Automatic
  name: tempo-product
  source: redhat-operators
  sourceNamespace: openshift-marketplace
EOF
----

.. Check the operator status by running the following command:
+
[source,terminal]
----
$ oc get csv -n openshift-tempo-operator
----

. Create a secret for your object storage bucket by running one of the following commands:

** To create a secret from a YAML file:
+
[source,terminal]
----
$ oc apply -f <secret_file>.yaml
----

** To create a secret from standard input:
+
[source,terminal]
----
$ oc apply -f - << EOF
<object_storage_secret>
EOF
----
+
--
include::snippets/distr-tracing-tempo-required-secret-parameters.adoc[]
--
+
--
include::snippets/distr-tracing-tempo-secret-example.adoc[]
--

. Create a project of your choice for the *TempoStack* instance that you will create in the next step:

** To create a project from standard input without metadata:
+
[source,terminal]
----
$ oc new-project <project_of_tempostack_instance>
----

** To create a project from standard input with metadata:

> [source,terminal]

Author: Apologies, I was checking the rendering myself earlier today and clearly missed this rendering error.
I've amended the PR just now.
Waiting for the preview to build.
Will ping you when the preview is ready so that you can verify it yourself.
